### ISSUE TITLE
Rotated Bounding Boxes

### ISSUE THREAD

**Issue by centazzo** [roboflow/rf-detr#56]
Is there a plan to support oriented bounding boxes in a future version? It would be extremely beneficial for applications in aerial imaging.

Thank you very much for your excellent work!

**Comment by isaacrob-roboflow** [comment]
there's no plan right now but noted, we'll keep in mind that there's interest! :) can you share a bit more about your use case? would segmentation solve?

**Comment by nmichlo** [comment]
Thanks @isaacrob-roboflow, are there plans for segmentation? (It would probably be more useful in the case I was considering).

**Comment by isaacrob-roboflow** [comment]
We plan to ship it but don't have an ETA!

**Comment by alecenta** [comment]
Thank you for you prompt reply @isaacrob-roboflow , yes there are quite a few use cases i can show you such as this image coming from DOTA dataset paper showcasing OBB superior clarity in particular in the context of aerial images (drones, satellites etc). 

![Image](https://github.com/user-attachments/assets/9e659d2f-75c0-4ffd-84ea-0dc960895886)

It became a topic also recently on Torchvision repo but as far as i know just a few steps were taken in that direction. Unfortunately segmentation can only partially provide a solution and the requirement of building segmentation masks to train a DL model are often unfeasible.

**Comment by isaacrob-roboflow** [comment]
I'm generally pretty pro obb, big downside in my mind is we don't have obb pretraining datasets. but it's possible to 'monkey patch' tooling for it into the model. supporting obb is a big discussion inside roboflow right now generally

your thoughts are super useful in informing our direction! 

what dataset format do you use to support obb training? can you talk a bit about your current pipeline?

**Comment by alecenta** [comment]
Currently working with (center x, center y, width, height, orientation) format. Not many fancy things in my current pipeline, I'm working with a two stage detector hence the "complex" part lies in computing rotated IOU efficiently both from memory and speed POV. Representing BBOXs in other format is also completely fine and the conversion to 4 corner coordinates for example is straightforward. As far as i know DOTA is the most comprehensive datasets with OBB and i saw that it is used also in YOLO models with OBBs.

**Comment by nmichlo** [comment]
If OBB support is added, will a 360deg option be added? There are cases where this is needed, but as far as I can tell things like yolo don’t support this only 180deg. Even things like predicting bboxes that extend outside the image.

In my opinion 4 point polygon is more useful than the standard rotated bounding boxes because you can handle perspective for objects. OBB are a then a subset of this in the 360deg case. Although not the 180deg case.

**Comment by isaacrob-roboflow** [comment]
@nmichlo do you know why those other frameworks only support 180deg?

**Comment by nmichlo** [comment]
I'm guessing here, but I think the main reason is if you want a minimal bounding box vs a truly oriented bounding box. 180° or 45° will get you a minimal one, 360 will get you a truly oriented one. Useful in different scenarios, could hurt performance if there isn't an option.

The frameworks also don't allow predicting bounding boxes that extend past the edge of the image, would be useful in some edge cases.

**Comment by nmichlo** [comment]
Thinking more about this, I definitely still think an `xyxyxyxy` (4 corners) implementation of OBB would be more useful than `xywhr` -- all cases in my previous comment could be covered with the former, with the latter being a subset. It then just becomes a dataset preparation task to handle truly oriented OBB vs minimal OBB.

Furthermore, xyxyxyxy is akin to a 4 point version of segmentation, so if you actually implement segmentation you've effectively generalised the training pipeline over OBB too, if you just limit the inputs to 4 points.

**Comment by TanvirKasir786** [comment]
@isaacrob-roboflow any updates on an ETA for when you will ship with OBB?

**Comment by vsingh-sereact** [comment]
@isaacrob-roboflow Any updates on supporting OBB?

**Comment by poilly54** [comment]
@isaacrob-roboflow  Any updates on supporting OBB?

**Comment by oguz-hanoglu** [comment]
Any updates on supporting OBB? YOLO has it. MMRotate has a better one. So the details of their implementation can serve as a basis maybe. Looking forward to see this great repo support this feature. Without OBB, simple 0 degree bboxes do not help much in our real world use case (satellite imagery analysis). Best regards.

**Comment by isaacrob** [comment]
I'm making a case for the platform team to build it out so that we have an excuse to build it here. The argument right now is that masking can handle this case. It would be very helpful if folks could give testimony as to why bbox or seg doesn't cover their cases :)

**Comment by oguz-hanoglu** [comment]
Thanks for pushing this forward @isaacrob . 
Happy to clarify why axis-aligned boxes or masks are not a real substitute for OBB in satellite imagery.

**Axis-aligned bboxes: (AABBs)**
With arbitrarily rotated objects, AABBs dramatically **increase background area**. For buildings, this can easily inflate footprint estimates by 2–3×. The enlarged boxes also create excessive overlap in dense scenes, leading to unwanted suppression during NMS and **lower recall.**

<img width="93" height="80" alt="Image" src="https://github.com/user-attachments/assets/342f52b7-0ce0-4dae-8354-7ab7d502de5f" />

**Masks:**
Masks can model orientation, but they are often **overkill** for mostly rectangular objects like buildings (higher annotation cost and inference overhead). More importantly, in dense urban areas where buildings are very close, instance masks **tend to merge** touching or nearly touching objects, making counting and per-instance area estimation unreliable. Additionally, mask outputs often have jagged boundaries and require post-processing to obtain clean footprints.

Note that OBB could only be replaced **by instance segmentation, not semantic segmentation**. Even then, it would be a heavier and less efficient alternative.

<img width="1089" height="697" alt="Image" src="https://github.com/user-attachments/assets/aa8f9296-bb33-498c-b537-202f5968b857" />

OBB provides a compact and geometrically correct representation that **preserves orientation** while remaining **lightweight**.

Happy to elaborate further if helpful.

**Comment by isaacrob** [comment]
Got it. It's worth noting that RF-DETR doesn't require NMS so the larger area causing unwanted suppression is a non issue. But for determining something like area for an object that is known to be rectangular ahead of time, certainly masking is slower for the same task.

**Comment by eorroot** [comment]
@isaacrob any updates on an ETA for when you will ship with OBB?

**Comment by jk4e** [comment]
+1 for OBB/Rotated Box!
+1 for Keypoints/Pose!

**Comment by janerikvw** [comment]
Are there updates on merging pullrequest #921 ? Great addition :)

**Comment by Rohphi** [comment]
> Are there updates on merging pullrequest [#921](https://github.com/roboflow/rf-detr/pull/921) ? Great addition :)

also would like to know.

**Comment by AmbroxMr** [comment]
> Are there updates on merging pullrequest [#921](https://github.com/roboflow/rf-detr/pull/921) ? Great addition :)

Interested as well on this one
